# Binary trees & Catalan number

Concevez un algorithme pour **énumérer tous les arbres binaires** d'une taille N donnée, où N correspond au nombre de noeuds (vertex) ; le N ne prend pas en compte les feuilles. Les noeuds feuilles sont représentés par ceci "(..)": une parenthèse ouverte '(', deux points '..' - représentant les deux feuilles- et une parenthèse fermée ')'.

Le programme final doit afficher la liste des arbres binaires de la taille spécifiée, un arbre par ligne. Pour N=3, il y a un total de 5 arbres binaires. L'output doit être une séquence de parenthèses et points, comme ceci:

    (((..).).)
    ((.(..)).)
    ((..)(..))
    (.((..).))
    (.(.(..)))
    
Implémentez et testez un algorithme récursif qui résout le problème. Donnez une solution pour N=[1,10]

Pour tester votre solution ("est-ce que la solution retourne le nombre correct d'arbres?"), vous devez lire, ~comprendre, implémenter et tester la formule qui calcule le nombre de Catalan pour un N donné:

C<sub>n</sub> = $\frac{(2n)!}{(n+1)!n!}$

Aide: https://fr.wikipedia.org/wiki/Nombre_de_Catalan

In [ ]:
import math
def catalan_nb(n: int) -> int:
    return math.comb(2 * n, n) // (n + 1)

In [4]:
def catalan_nb(n: int):

    if n < 0 :
        return exit
  
    elif n <= 1:
        return 1 
    
    # Recursive Case
    # catalan_nb(n) is the sum of catalan_nb(i)*catalan_nb(n-i-1)
    res = 0
    for i in range(n):
        res += catalan_nb(i) * catalan_nb(n-i-1) 
 
    return res


for i in range(1, 11, 1):  
    print(catalan_nb(i), end=' ')


1 2 5 14 42 132 429 1430 4862 16796 

In [17]:
assert catalan_nb(3) == 5
assert catalan_nb(5) == 42
assert catalan_nb(8) == 1430

In [5]:
def get_all_B_Trees(n):
    
    # Base case

    if n == 0:
        return [ "." ]
    
    # Recursive case
    
    else:

        trees = []

        for left_nodes in range(n):
            for left_tree in get_all_B_Trees(left_nodes):
                for right_tree in get_all_B_Trees(n - left_nodes - 1):
                    tree = "(" + left_tree + right_tree + ")" 
                    trees.append(tree)
        
        return trees
    
for tree in get_all_B_Trees(3):
    print(tree)

(.(.(..)))
(.((..).))
((..)(..))
((.(..)).)
(((..).).)


Ecrivez ci-dessous votre réponse pour N=[1,10]

For N = 1 :
catalan_nb = 1 

For N = 2 :
catalan_nb = 2 

For N = 3 :
catalan_nb = 5

For N = 4 :
catalan_nb = 14

For N = 5 :
catalan_nb = 42

For N = 6 :
catalan_nb = 132

For N = 7 :
catalan_nb = 429

For N = 8 :
catalan_nb = 1430

For N = 9 :
catalan_nb = 4862

For N = 10 :
catalan_nb = 16796

### Explications

**catalan_nb algorithm #1**
The first one, with the math module, is closer to the fraction C<sub>n</sub> = $\frac{(2n)!}{(n+1)!n!}$, but isn't recursive

**catalan_nb algorithm #2**

1. Pre-base case : Because we're considering all n >= 0, if n < 0, exit the function
2. Base case : The catalan number(s) for 0 and 1 are 1. So if n == 0 or n == 1, return 1.
3. Recursive case : For n > 1, the catalan_nb(n) is the sum of catalan_nb(i)*catalan_nb(n-i-1).

Pour l'algorithme Catalan, on commence par établir que les chiffres négatifs ne sont pas acceptables. Ensuite lorsque le nombre est entre 0 et 1, mon résultat sera de toute manière 1. Et finalement notre cas récursif, la fonction Catalan est la sommme du nombre de noeuds multipliés par (n- i - 1)

**get_all_B_Trees algorithm**

1. Base case : if n == 0, then there's still no leaves, only the beginning node, which is ".".
2. Recursive case : the variable "trees" represent the list of binary trees according to n. "(..)" represents a tree. The first parameter is the left sub-tree and the next one is the right sub-tree.

Quand on a aucun noeud c'est N=0, on a une feuille désignée par "." (structure fondamentale), c'est le point du début de récursivité. Pour le développement, "(..)" représente un arbre la dispersion se fait de manière N-1 noeuds vu que un noeuds c'est une racine d'arbre. Donc pour chaque N on essaie toutes les options possibles de sous-arbres de gauche (left) puis à la droite (right). Exemple : N=3 on a 2 sorties à repartir en sous arbres, nous donnant donc un résultat de 5.

Qu'elle est la complexité en espace/temps de l'algorithme ? Une approximation asymptotique est recommandée. Vous aurez besoin de la formule de Stirling pour cela : $n ! \approx \sqrt{2 \pi n}\left(\frac{n}{\mathrm{e}}\right)^n$

O($n!$) (asymptotiquement équivalente à O($n^n$))

Lorsque l'algorithme implique des factorielles comme la formulation de Stirling le propose, alors on peut faire une séparation et vérifier les asymptotes correctement, avec une complexité générale de O($n!$). Une factorielle est une multiplication de O(n) (la complexité en espace est donc O(n), profondeur linéaire de l'appel récursif). Pour la partie $(\frac{n}{\mathrm{e}})^n$ il s'agit a priori d'une partie de complexité O($n^n$)

En conclusion, l'algorithme `get_all_B_Trees` possède une complexite en temps exponentielle. Cela est dû à la nature récursive de l’algorithme et à la croissance exponentielle du nombre d’arbres binaires possibles à mesure que l’entrée `n` augmente.